nexus/src/db/model.rs

```diff
@@ -1741,18 +1763,61 @@ impl Into<external::VpcFirewallRule> for VpcFirewallRule {
     }
 }
 
-#[derive(Queryable, Insertable, Clone, Debug, Resource)]
+/// A not fully constructed NetworkInterface. It may not yet have an IP
+/// address allocated.
+#[derive(Clone, Debug)]
+pub struct IncompleteNetworkInterface {
+    pub identity: NetworkInterfaceIdentity,
+
+    pub instance_id: Uuid,
+    pub vpc_id: Uuid,
+    pub subnet: VpcSubnet,
+    pub mac: MacAddr,
+    pub ip: Option<std::net::IpAddr>,
```

Comment:
Why have `ip` on here as optional rather than leaving it off entirely? From what I can tell it looks like it's unused (so far).

Reply:
I believe it's used in the creation logic to condition on whether to generate an IP or to use a given one. I had included it based on the RFD suggesting that VMs may be created with specific IPs. Fine removing it though if you'd rather

Reply:
I see, so once there's an endpoint to create a network interface, presumably that would take an optional IP, and this field will be used that way. Could it make sense to move the logic that matches on IP presence up a level out of the datastore into `nexus.rs`, so that the optional IP would be on the `NetworkInterfaceCreate` params, but then you wouldn't need to represent it at the model level at all, i.e., by the time you make a `db::model::NetworkInterface`, you always already have the IP, whether it was passed in or you had to generate it.

Reply:
Actually before you bother answering that, I see the answer is probably in the `AllocateIpQuery` since we're doing it all in one query. Let me take a look at that.

Reply:
The current complication is that we don't have atomic allocation, so we'll need a retry loop with that. Though the current thing isn't actually atomic, it just narrows the window for a race, and should probably get a retry loop too. Longer term it might be useful to decouple the IP allocation from the network_interface table to allow us to exclusively allocate an IP before constructing the network_interface so that the network interface creation itself doesn't have to deal with retrying from that

Reply:
I see. Might be nice to add a TODO someplace about the ideal end state even if we're not sure when we'll get there. Maybe in the datastore function since that's the part that would change the most. Or track it in an issue.